[Docs][Qt][Tests][Wallet] Adding Seed phrases

panleone opened the pull request:
Coauthored with @Duddino 
The aim of this pull request is adding seed phrases support with the BIP39 standard. 

What I have done:
- Added a function to generate a random 24 words seed phrase
- Added a function to convert the seed phrase in the corresponding seed
- Added a function to check if a seed phrase is valid
- Implemented a qt interface
- Implemented -seedphrase="words..." command on pivxd 
- wrote test units and functional tests
- Implemented support for multi languages

EDIT:
we should discuss external compatibility in another PR

Liquid369 commented:
Just getting into this now and as the start goes this is an concept uTACK!

A few things along the way for the basics of this before we get into the next TODO's

src/wallet/bip39_english.txt what if we do a JSON CPP file? IMO, it would look much cleaner. Do you like that idea?

In src/wallet.h we have the following enum class for "wallet versioning"

```/** (client) version numbers for particular wallet features */
enum WalletFeature {
    FEATURE_BASE = 10500, // the earliest version new wallets supports (only useful for getinfo's clientversion output)

    FEATURE_WALLETCRYPT = 40000, // wallet encryption
    FEATURE_COMPRPUBKEY = 60000, // compressed public keys

    FEATURE_PRE_PIVX = 61000, // inherited version..

    // The following features were implemented in BTC but not in our wallet, we can simply skip them.
    // FEATURE_HD = 130000,  Hierarchical key derivation after BIP32 (HD Wallet)
    // FEATURE_HD_SPLIT = 139900, // Wallet with HD chain split (change outputs will use m/0'/1'/k)

    FEATURE_PRE_SPLIT_KEYPOOL = 169900, // Upgraded to HD SPLIT and can have a pre-split keypool

    FEATURE_SAPLING = 170000, // Upgraded to Saplings key manager.

    FEATURE_LATEST = FEATURE_SAPLING
};
```

We should add a new ```FEATURE_WORDLIST = 180000``` as a possible idea.
Giving ourselves a differentiation in the working of this and to not confuse others. 
Alongside having several "wallet update/upgrade functions" that we would want to account for in checking if our wallet is at the current installed version or if the user would like to update to a seed phrase
There is a starting param ```-upgradewallet```  in init.cpp line 31/93/94

For testing purposes, if we add RPC for validating/creating seeds would be helpful vs the current validation of a hard coded seed and make sure it can be repeated.

Partial review, before digging further I think these are some possible feature additions for this current portion to iron out and we can look into the other options. 

Will update later here or in successive comment after testing!
Great start!

panleone commented:
This is the UI that I created
![newUI](https://user-images.githubusercontent.com/22001825/206911901-b40df08f-fea0-4f5c-8e2c-5ed85cf8ff79.png)

panleone commented:
With this commit the QT part should be finished and functional... here a picture of the import seedphrase UI
![import](https://user-images.githubusercontent.com/22001825/207629327-4d933051-9737-4dda-adb3-fb8797314ff5.png)

panleone commented:
With this commit seedphrases are fully supported on pivxd as well.
with the arg:
`./pivxd -seedphrase="hi this is a seed phrase..."`
A seed phrase can be imported on pivxd and in particular this works only when the wallet is opened for the first time.
if instead the `-seedphrase` arg is not specified a new random seedphrase is generated and printed on console

PeterL73 commented:
Nice feature. Looking at an other project that implemented this in a different way. 
https://github.com/Veil-Project/veil/tree/83db278ddb89e3096ca5b9de10965b7ad8f8e522/src/veil/mnemonic
Might be nice to consider this (and gain other languages).

panleone commented:
> Nice feature. Looking at an other project that implemented this in a different way. https://github.com/Veil-Project/veil/tree/83db278ddb89e3096ca5b9de10965b7ad8f8e522/src/veil/mnemonic Might be nice to consider this (and gain other languages).

Agree... the dictionary is probably better than making a file for each language

Fuzzbawls commented:
Drop commit d6c6a87745a0147d7eafd284ea2eb53f19398d01. It only serves to complicate the review process. Don't apply style fixes broadly to completely unrelated files, only to NEW code you're submitting or are directly changing.

Fuzzbawls commented:
Use the `clang-format-diff.py` script to do style fixups to only code you've added/changed. for this PR, the command would be:
```bash
git diff -U0 c130d4c5591a5d27a3b597e3882234ef9da9d3e3^..b3d95480af008d607cc1ce309da46aaf1bab4d74 | ./contrib/devtools/clang-format-diff.py -p1 -i -v
```

Duddino reviewed (commented):
This is looking really good!
I have two small nitpicks: the import screen text could be larger, and when it fails to validate the seedphrase, it could tell which word is not in the dictionary / if the problem is coming from the checksum

panleone commented:
> This is looking really good! I have two small nitpicks: the import screen text could be larger, and when it fails to validate the seedphrase, it could tell which word is not in the dictionary / if the problem is coming from the checksum

Hi Duddino, thanks for the review, I improved the error handling as you suggested, for the text size I think that they are big enough, but let's wait for more opinions

floki008 reviewed (approved):
tACK: 
Tested on linux ubuntu ryzen 7 5800x, 16GB ram, ssd internal and external 1TB
all work no bug

Liquid369 reviewed (commented):
Alright looking very good so far with one trivial change!
![Screenshot 2023-01-26 at 10 55 00 AM](https://user-images.githubusercontent.com/45834289/214905243-fb5d8486-7a16-447d-bc90-099222df1c23.png)
The seed phrase page looks exceptionally small in regard to my screen and trying to read it. We do want users to be able to see it and record it properly.
With that also being said, how about a function/button to copy the entire seed for the user if they're going to copy it into a notepad and then write it down later
Trivial in most cases but figured I would mention this.

I like the flow because if you have a wallet you are not forced to change anything. 

When using the daemon for people or services we want to give them the opportunity to create a seed also using pivxd upon first start/creation, but we can tackle this in a subsequent PR, masternodes are not a huge necessity to have a seed phrase for. This would be for particular services mainly anyways

panleone commented:
> Alright looking very good so far with one trivial change! ![Screenshot 2023-01-26 at 10 55 00 AM](https://user-images.githubusercontent.com/45834289/214905243-fb5d8486-7a16-447d-bc90-099222df1c23.png) The seed phrase page looks exceptionally small in regard to my screen and trying to read it. We do want users to be able to see it and record it properly. With that also being said, how about a function/button to copy the entire seed for the user if they're going to copy it into a notepad and then write it down later Trivial in most cases but figured I would mention this.
> 
> I like the flow because if you have a wallet you are not forced to change anything.
> 
> When using the daemon for people or services we want to give them the opportunity to create a seed also using pivxd upon first start/creation, but we can tackle this in a subsequent PR, masternodes are not a huge necessity to have a seed phrase for. This would be for particular services mainly anyways

Hi Liquid thanks for the review! As I already said on discord seedphrases are already implemented on pivxd (check the PR description). I didn't expect the GUI to look so bad on your computer, I will try to fix in those days

panleone commented:
> Sorry for jumping in this late in the conversation, but I am curious to know if you can import a bip39 seedphrase + password into the core wallet? Or is there a way to import the bip32 root key? Both should achieve the same result: output all the dirived addresses. If this is possible, is there any documentation? I have searched without success.
> 
> If this is an upcoming feature, when should users expect it to be implemented in the core program?

You will be able to import/generate bip39 seedphrases once this PR is merged and the new wallet version comes out (probably with 6.0 I'd say). The feature is basically completed/working I only need to  design a better UI interface because it is too small. Also I'm currently working on DMNs (and I'm a Qt noob) so I don't know when I will finish.

For the bip32 AFAIK there is no "easy" way to import the master key or seed (as it should be since it is very easy to make mistake with a random sequence of bytes)

panleone commented:
> @panleone, is there a way to do this through the console right now?

At the moment (v5.5) the most you can do is choosing the hd seed with the rpc command: sethdseed, but beware that some action (for example locking the wallet with a password) will change the seed again.

panleone commented:
> @panleone I'm not seeing this anywhere in [the milestone section](https://github.com/PIVX-Project/PIVX/milestone/34), are you certain this will be making its way into the v6 release?

If other devs agree to release it into v6 yes

panleone commented:
Current status of this PR: Finally after 2 months I fixed all suggested changes, i.e. added support for multi languages seedphrases (at the moment there is italian and english) and improved the qt in general with the help of Liquid.

Liquid369 reviewed (approved):
tACK 24b6009f3980e2428e99811a5c3ea8c312e0ac14
Much better since the recent Qt fixes, I think we are ready now!

Fuzzbawls commented:
This PR should be split up into at least 2 distinct PRs: first PR would implement the BIP39 functionality on the network level, part 2 would add GUI functionality.

Also, including multi-language hard-coded string translations directly in the source tree needs more consideration and evaluation.
